### PR TITLE
use YT player API to force autoplay, resolving double-tap bug

### PR DESCRIPTION
**Pull request by paulirish**
hopefully fixing  #6
inspired by #90

See https://github.com/paulirish/lite-youtube-embed/issues/6#issuecomment-555135068 for some context and details

Thanks @dantovein for kicking off this approach. you'll see a lot of similarities to #90.

I originally wanted to avoid loading this JS. (it's 300kb), but since the iframe JS is already 1MB, it seems fairly reasonable to add another 30%. :/ 
(god how i wish the typical iframe payload was smaller. alas.)

---------

Seeking feedback on this approach! 

Also happy for folks to test this. It seems to fix the double-tap/non-autoplay bug 95% of the time.  Test pages: https://lite-youtube-embed-git-ytapi-paulirish.vercel.app/

### CODE REVIEW

File: src/lite-yt-embed.js [hunk 1]
```diff
@@ -52,6 +52,13 @@ class LiteYTEmbed extends HTMLElement {
         // TODO: In the future we could be like amp-youtube and silently swap in the iframe during idle time
         //   We'd want to only do this for in-viewport or near-viewport ones: https://github.com/ampproject/amphtml/pull/5003
         this.addEventListener('click', this.addIframe);
+
+
+        // Chrome & Edge have no problem with the basic YouTube Embed with ?autoplay=1
+        // However Safari and Firefox do not successfully track the user gesture of clicking through the creation/loading of the iframe,
+        // so they don't autoplay automatically. Instead we must load an additional 300KB (ungz) of JS for the YT Player API
+        // TODO: chrome android seems to also need this
+        this.needsYTApiForAutoplay = navigator.vendor.includes('Apple') || navigator.userAgent.includes('Firefox');
```
Comment (paulirish) [inline src/lite-yt-embed.js:61]: i figure ill have to change this to be some other sniff. don't think there's a good "feature detect" that's equivalent to failing to successfully autoplay. so yeah.. we'll do a UA sniff of some sort.  

but this'll need to also include chrome android as well. and perhaps more.   

some testing required:

#### has problems with autoplay on https://lite-youtube-embed-git-master-paulirish.vercel.app/

ios safari
desktop safari
firefox android
chrome android

#### no problems

desktop chrome

#### ???

firefox desktop
edge desktop and mobile

File: src/lite-yt-embed.js [hunk 2]
```diff
@@ -95,12 +102,53 @@ class LiteYTEmbed extends HTMLElement {
...
+    }
+
+    async addYTPlayerIframe(params) {
+        this.fetchYTPlayerApi();
+        await this.ytApiPromise;
+
+        const videoPlaceholderEl = document.createElement('div')
+        this.append(videoPlaceholderEl);
+
+        const paramsObj = Object.fromEntries(params.entries());
```
Comment (paulirish) [inline src/lite-yt-embed.js:127]: im a little scared this newish feature will pose problems with various people's tsc/acorn/estree versions, if they're too old. but hopefully its fine.

File: src/lite-yt-embed.js [hunk 3]
```diff
@@ -95,12 +102,53 @@ class LiteYTEmbed extends HTMLElement {
...
+        });
+    }
+
+    async addIframe(){
         if (this.classList.contains('lyt-activated')) return;
         this.classList.add('lyt-activated');
 
         const params = new URLSearchParams(this.getAttribute('params') || []);
         params.append('autoplay', '1');
+        params.append('playsinline', '1');
```
Comment (paulirish) [inline src/lite-yt-embed.js:147]: https://developers.google.com/youtube/player_parameters#april-27,-2021

File: src/lite-yt-embed.js [hunk 4]
```diff
@@ -95,12 +102,53 @@ class LiteYTEmbed extends HTMLElement {
         LiteYTEmbed.preconnected = true;
     }
 
-    addIframe() {
+    fetchYTPlayerApi() {
```
Comment (paulirish) [inline src/lite-yt-embed.js:105]: alsoo wondering if there's a lazy way to attach... https://github.com/paulirish/lite-youtube-embed/issues/82  hmmm!

File: src/lite-yt-embed.js (hunk 2 above)
Comment (codepuncher) [reply src/lite-yt-embed.js:127]:
```suggestion
        const paramsObj = params.entries();
```
I believe using `Object.fromEntries` is redundant here anyway.

**Comment by paulirish** [comment]
From https://github.com/paulirish/lite-youtube-embed/issues/6#issuecomment-1242944293 ..

> Thanks for the YT API merge.
> 
> It’s now working with one click/tap on almost all browsers except ipadOS Safari (and most likely iOS, but I don’t have an iPhone to test). Tested and worked well on Firefox Desktop, Chrome Desktop, Safari Desktop and Chrome mobile.

I don't have an ipad sadly.. anyone wanna pair up with me to sort this out?

**Comment by simevidas** [comment]
https://lite-youtube-embed-git-ytapi-paulirish.vercel.app/

On iPhone, I have to tap the play button twice to play the video.

**Comment by nucliweb** [comment]
Ey @paulirish, you can test it with iOS Simulator

If you add the event `onStateChange` we can see the states

```js
new YT.Player(videoPlaceholderEl, {
    width: '100%',
    videoId: this.videoId,
    playerVars: paramsObj,
    events: {
        'onReady': event => {
            event.target.playVideo();
        },
        'onStateChange': event => console.log(event) 👈
    }
});
```

## Sreenshots

**Pull request merged**